### DISCUSSION TITLE
Could System.ServiceModel.Duplex act as foundation for CoreWCF

### DISCUSSION THREAD

**Discussion by Danielku15** [CoreWCF/CoreWCF#536] [Ideas]
One of the design decisions for CoreWCF was to not pull the client side WCF packages as foundation. I can understand these concerns especially as they only forward to the .net Framework System.ServiceModel which breaks the migration scenarios intended with CoreWCF. 

Nevertheless I was wondering if it is really worth the effort maintaining so much of the WCF foundation two times. CoreWCF is quasi independent (with some compat layer to System.ServiceModel attributes) but maintains a lot of things separately from the System.ServiceModel. 

While trying to use CoreWCF and dotnet/wcf for our .net core transition I was repeatingly thinking of this idea: Actually dotnet/wcf (client side) supports Duplex Channels and with the callback contracts, actually the client can become the server of two connected services. All the heavy lifting of the message encoding/decoding, the dispaching messages etc. it should all be there already. What is missing, is the "hosting" aspects of CoreWCF. 

I see a lot of opportunity to have a quite nice shared foundation for the server and client parts of WCF without limiting CoreWCF for extensions and not spoiling the client parts with the server needs. Shared components could cover: 

- The contract attributes and various common data structures. 
- Base classes for the Bindings and other configuration related aspects (open for extension by client or server parts)
- All internals needed for operation dispatching (messaging classes)
- Base Message encoders/decoders and implementations for some common formats (http, nettcp). 

The dotnet/wcf usage of these shared components would add client specifics on top like the ChannelFactories.
The CoreWCF usage of these shared components would add the hosting specifics on top like the ASP.net core integration into Kestrel, hosting specific binding settings (e.g.`NetTcpBinding.MaxConnections`), WSDL generation,..

Any additional channels like gRPC, Named Pipes etc. likely anyhow also should get client side support to really bring future value. Backwards compatibility is good, but being realistic, if we add WCF over gRPC server capabilities, people will expect to get a WCF over gRPC client. 

I was wondering if you guys also had such an idea in mind already. I found it quite exciting and was close to attempt faking a WCF Server by having a WCF Duplex Client as backbone.

**Comment by mconnew** [comment]
Unfortunately that won't work. The common code needs to know about server and client concepts so would pollute each project with concepts from the other. It would also result in preventing CoreWCF from deviating from WCF's api's in many places. Take for example the class Binding which all Binding's derive from. That has CanBuildChannelFactory and BuildChannelFactory methods, which returns instances of ChannelFactory. Having those methods in CoreWCF would mean it would need to pull in those classes from the client. The initial thought is that's okay, just put those classes into the common package as it's a small price. But then that pulls in ClientCredentials. Which pulls in a lot of client side only implementation. It's a thread that you keep pulling on and pulling on until you end up with only a small amount of the implementation which can be in separate client packages. This causes problems as you now add dependency to ASP.NET Core on the client side. The WCF Core client is used on Android and iOS in the latest Xamarin versions, and this wouldn't be possible. It would also prevent WCF Core client from being usable in Blazor (I don't know if it works there today, but in theory it should be doable without too much work if it doesn't yet).

You then have the same problem the other way around, but this time it's actually worse. We've got rid of the IChannelListener interfaces and related classes in favor of a push based middleware model similar to ASP.NET Core's middleware. This causes a problem with the contracts. For the client implementation, when running on .NET Framework there are facade's distributed in the nuget package which forward the types to System.ServiceModel.dll in .NET Framework. This means we can't change the shape of any api's. We can add new api's for .NET Core only, but can't modify existing ones. This would have added a lot of overhead in trying to add new api's, and would have made development of the server side of code a lot slower.

I mentioned the move from using APM async to the push based middleware model. I did a massive refactor/rewrite of the code from WCF to CoreWCF to get rid of APM. To say that WCF abuses APM to its limit is an understatement (my understanding is the original WCF development team invented the concept of APM in .NET). The code is incredibly optimized, but the result is that you can't casually work on the code base. The ramp up time on being able to make meaningful changes beyond trivial is measured in months of full time work on the code base. While I've got rid of as much APM code as I can on the client side, it's still exposed on the public api's, and this requires maintaining some difficult to grok code to make out of box extensibility work while using Task's internally. Having a common core would pollute CoreWCF with this requirement. One of my aims with CoreWCF was to make it so that 90% of the code was something a mid level developer would be able to follow and understand within a couple of hours with a debugger. The fact that CoreWCF gets more community contributions in a month than the WCF Core client has had in the projects entire lifetime is down to this fact.  

There's also the problem of forcing concurrent implementations to be developed in both projects by having things common. For example, the reliable sessions feature in the client was relatively easy to port, it took me about 2 days to do the core porting work. For CoreWCF, it's going to take a lot more time. There are some design issues with how it's implemented in WCF which can cause some problems which prevent a straight port. Making this code common would have delayed release of the client implementation.  

As for gRPC support, you're correct about requiring client and server to be released together. In the next few months I'm going to see how that model works. On the client side I'll be working on supporting named pipes pretty soon. But named pipes doesn't work on Linux/Mac, but unix domain sockets do. ASP.NET Core doesn't support named pipes natively, but it does support UDS. I'm going to implement UDS support for CoreWCF to have an option on Linux/Mac, and the WCF Core client will also need an implementation. But then you have .NET Framework, it would be ideal to have support there so everyone can talk to everyone else. I'm going to attempt to find a clean way to have named pipes and UDS support everywhere with as little duplication of code as possible. We'll see how well that works.  

Talking more about gRPC, this is another area where having common code would hinder things. WCF/CoreWCF is very SOAP opinionated in its object model. For CoreWCF v2 I want to de-opinionate CoreWCF. What I think this will look like is the current Message class will be given a base class, maybe called MessageBase which has common (probably generic) methods and properties for all protocols. Then the Message class will derive from that and be SOAP opinionated. Then we could add a GrpcMessage class which is gRPC opinionated. The rest of the CoreWCF infrastructure will then reference the unopinionated MessageBase class. For example today `IDispatchMessageFormatter.DeserializerRequest` takes a Message and an array of objects and populates the array with the parameters for the operation. This would be changed to take a MessageBase and CoreWCF would just pass the MessageBase object to it. It wouldn't care that MessageBase is actually a GrpcMessage, just that the array of objects gets populated. This refactoring wouldn't be possible if all of this code was common with the client. On the client side, the same tricks that are used with WebHttpBinding will need to be used for gRPC, that is wrap the real data in a dummy Message object and provide extensibility implementations which pull out the relevant parts, so basically make a gRPC request look like a SOAP Message.

Basically it's really complicated and there are many problems with this approach. I did actually try. I gave up after a weeks worth of work.

**Reply by Kralizek** [reply]
Does this mean that we can never have support for duplex, regardless of its implementation?

**Reply by mconnew** [reply]
We support duplex in CoreWCF today. This was just about making a common package which is used by the client and the server. If it were possible, it would reduce code duplication, and allow using a single namespace for both projects. As it stands right now, some extensibility points in WCF are applicable to client and server, so if you have an implementation which you use for both, you need to compile your code twice, once for the client using System.ServiceModel namespace, and once for the server using CoreWCF namespace. One example of this is `IEndpointBehavior` can be used for client and server.

**Reply by Kralizek** [reply]
I understand. Thanks for the clarification!
